### PR TITLE
[IMP] util/models: ensure execution of necessary updates

### PR DESCRIPTION
**Pull request by hpr-odoo**
- Executing update query even if there is no related record to Update, it's useless
- well, thinking about it due to facing memory error

- Traceback:
```
2024-03-08 04:26:19,433 32 INFO 1401619_org_14.0 odoo.modules.migration: module sales_team: Running migration [>saas~13.3.1.0] pre-migrate
2024-03-08 04:26:19,485 32 INFO 1401619_org_14.0 odoo.upgrade.util.pg: Dropping constraint crm_tag_name_uniq on table crm_tag
2024-03-08 04:26:19,487 32 INFO 1401619_org_14.0 odoo.upgrade.util.pg: Renaming FK 'crm_lead_tag_create_uid_fkey' to 'crm_tag_create_uid_fkey'
2024-03-08 04:26:19,488 32 INFO 1401619_org_14.0 odoo.upgrade.util.pg: Renaming FK 'crm_lead_tag_team_id_fkey' to 'crm_tag_team_id_fkey'
2024-03-08 04:26:19,488 32 INFO 1401619_org_14.0 odoo.upgrade.util.pg: Renaming FK 'crm_lead_tag_write_uid_fkey' to 'crm_tag_write_uid_fkey'
2024-03-08 04:27:41,769 32 WARNING 1401619_org_14.0 odoo.modules.loading: Transient module states were reset
2024-03-08 04:27:41,769 32 ERROR 1401619_org_14.0 odoo.modules.registry: Failed to load registry
Traceback (most recent call last):
  File "/home/odoo/src/odoo/14.0/odoo/modules/registry.py", line 89, in new
    odoo.modules.load_modules(registry._db, force_demo, status, update_module)
  File "/home/odoo/src/odoo/14.0/odoo/modules/loading.py", line 457, in load_modules
    force, status, report, loaded_modules, update_module, models_to_check)
  File "/home/odoo/src/odoo/14.0/odoo/modules/loading.py", line 349, in load_marked_modules
    perform_checks=perform_checks, models_to_check=models_to_check
  File "/home/odoo/src/odoo/14.0/odoo/modules/loading.py", line 174, in load_module_graph
    migrations.migrate_module(package, 'pre')
  File "/home/odoo/src/odoo/14.0/odoo/modules/migration.py", line 186, in migrate_module
    migrate(self.cr, installed_version)
  File "/tmp/tmpv851tzf_/migrations/sales_team/saas~13.3.1.0/pre-migrate.py", line 8, in migrate
    util.rename_model(cr, "crm.lead.tag", "crm.tag")
  File "/tmp/tmpv851tzf_/migrations/util/models.py", line 243, in rename_model
    explode_execute(cr, query, table=table)
  File "/tmp/tmpv851tzf_/migrations/util/pg.py", line 227, in explode_execute
    logger=logger,
  File "/tmp/tmpv851tzf_/migrations/util/pg.py", line 131, in parallel_execute
    return parallel_execute_impl(cr, queries, logger=_logger)
  File "/tmp/tmpv851tzf_/migrations/util/pg.py", line 112, in _parallel_execute_threaded
    executor.map(execute, queries),
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 575, in map
    fs = [self.submit(fn, *args) for args in zip(*iterables)]
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 575, in <listcomp>
    fs = [self.submit(fn, *args) for args in zip(*iterables)]
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 120, in submit
    w = _WorkItem(f, fn, args, kwargs)
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 46, in __init__
    self.future = future
MemoryError
```

- sql data:
```
1401619_org_14.0=> SELECT model, name
          FROM ir_model_fields
         WHERE ttype='reference';
          model          |     name
-------------------------+--------------
 theme.ir.ui.view        | inherit_id
 marketing.campaign.test | resource_ref
 sale.metric             | record_id
 ir.actions.server       | copy_object
 ir.ui.menu              | action
 ir.server.object.lines  | resource_ref
 crm.lead                | ref2
 marketing.participant   | resource_ref
 calendar.attendee       | ref
 crm.lead                | ref
 connector.checkpoint    | backend_id
 connector.checkpoint    | record
 corrective.action       | ref
 crm.event.log           | source
 avatax.transaction      | source
 mail.template.preview   | resource_ref
 rating.rating           | resource_ref
 rating.rating           | parent_ref
(18 rows)

1401619_org_14.0=> SELECT c.relname, a.attname, COALESCE(bt.typname, t.typname) AS udt_name,
               NOT (a.attnotnull OR t.typtype = 'd' AND t.typnotnull) AS is_nullable,
               (   c.relkind IN ('r','p','v','f')
               AND pg_column_is_updatable(c.oid::regclass, a.attnum, false)
               ) AS is_updatable
          FROM pg_attribute a
          JOIN pg_class c
            ON a.attrelid = c.oid
          JOIN pg_type t
            ON a.atttypid = t.oid
     LEFT JOIN pg_type bt
            ON t.typtype = 'd'
           AND t.typbasetype = bt.oid
         WHERE (c.relname = 'marketing_campaign_test' AND a.attname = 'resource_ref')
            OR (c.relname = 'sale_metric' AND a.attname = 'record_id')
            OR (c.relname = 'ir_actions_server' AND a.attname = 'copy_object')
            OR (c.relname = 'ir_ui_menu' AND a.attname = 'action')
            OR (c.relname = 'ir_server_object_lines' AND a.attname = 'resource_ref')
            OR (c.relname = 'crm_lead' AND a.attname = 'ref2')
            OR (c.relname = 'marketing_participant' AND a.attname = 'resource_ref')
            OR (c.relname = 'calendar_attendee' AND a.attname = 'ref')
            OR (c.relname = 'crm_lead' AND a.attname = 'ref')
            OR (c.relname = 'connector_checkpoint' AND a.attname = 'backend_id')
            OR (c.relname = 'connector_checkpoint' AND a.attname = 'record')
            OR (c.relname = 'corrective_action' AND a.attname = 'ref')
            OR (c.relname = 'crm_event_log' AND a.attname = 'source')
            OR (c.relname = 'avatax_transaction' AND a.attname = 'source')
            OR (c.relname = 'mail_template_preview' AND a.attname = 'resource_ref')
            OR (c.relname = 'rating_rating' AND a.attname = 'resource_ref')
            OR (c.relname = 'rating_rating' AND a.attname = 'parent_ref')
;
        relname        |   attname    | udt_name | is_nullable | is_updatable
-----------------------+--------------+----------+-------------+--------------
 mail_template_preview | resource_ref | varchar  | t           | t
 ir_ui_menu            | action       | varchar  | t           | t
 calendar_attendee     | ref          | varchar  | t           | t
 corrective_action     | ref          | varchar  | t           | t
 avatax_transaction    | source       | varchar  | f           | t
 crm_lead              | ref2         | varchar  | t           | t
 crm_lead              | ref          | varchar  | t           | t
 connector_checkpoint  | backend_id   | varchar  | f           | t
 crm_event_log         | source       | varchar  | f           | t
 sale_metric           | record_id    | varchar  | t           | t
(10 rows)

1401619_org_14.0=> SELECT 1 FROM mail_template_preview WHERE resource_ref LIKE 'crm.lead.tag,%' LIMIT 1;
 ?column?
----------
(0 rows)

1401619_org_14.0=> SELECT 1 FROM ir_ui_menu WHERE action LIKE 'crm.lead.tag,%' LIMIT 1;
 ?column?
----------
(0 rows)

1401619_org_14.0=> SELECT 1 FROM calendar_attendee WHERE ref LIKE 'crm.lead.tag,%' LIMIT 1;
 ?column?
----------
(0 rows)

1401619_org_14.0=> SELECT 1 FROM corrective_action WHERE ref LIKE 'crm.lead.tag,%' LIMIT 1;
 ?column?
----------
(0 rows)

1401619_org_14.0=> SELECT 1 FROM avatax_transaction WHERE source LIKE 'crm.lead.tag,%' LIMIT 1;
 ?column?
----------
(0 rows)

1401619_org_14.0=> SELECT 1 FROM crm_lead WHERE ref2 LIKE 'crm.lead.tag,%' LIMIT 1;
 ?column?
----------
(0 rows)

1401619_org_14.0=> SELECT 1 FROM crm_lead WHERE ref LIKE 'crm.lead.tag,%' LIMIT 1;
 ?column?
----------
(0 rows)

1401619_org_14.0=> SELECT 1 FROM connector_checkpoint WHERE backend_id LIKE 'crm.lead.tag,%' LIMIT 1;
 ?column?
----------
(0 rows)

1401619_org_14.0=> SELECT 1 FROM crm_event_log WHERE source LIKE 'crm.lead.tag,%' LIMIT 1;
 ?column?
----------
(0 rows)

1401619_org_14.0=> SELECT 1 FROM sale_metric WHERE record_id LIKE 'crm.lead.tag,%' LIMIT 1;
 ?column?
----------
(0 rows)
```

- In actual issue is about their data
```
1401619_org_14.0=> select count(id) from sale_metric;
 count
--------
 577552
(1 row)

1401619_org_14.0=> select max(id) from sale_metric;
     max
-------------
 58406524824
(1 row)
```

- UPG-1401619
- OPW-3766454

### CODE REVIEW

**Comment by hpr-odoo** [comment]
Hello @odoo/upgrade Team, Could you please review this PR?
Matt is fail but not related to this change.

**Review by aj-fuentes — COMMENTED** [review]
I'm testing an old idea to solve this and other issues. Some comments for your current patch in the meantime.

File: src/util/models.py [hunk 1]
```diff
@@ -233,14 +233,20 @@ def rename_model(cr, old, new, rename_table=True):
...
+                    table=table, column=column, old=old
+                )
+            )
+            if cr.rowcount:
+                query = cr.mogrify(
+                    """
+                        UPDATE "{table}"
+                           SET "{column}"='{new}' || substring("{column}" FROM '%#",%#"' FOR '#')
+                         WHERE "{column}" LIKE '{old},%'
+                """.format(table=table, column=column, new=new, old=old)
```
[the remark this answers is not in the document]
Comment (hpr-odoo) [reply src/util/models.py]: Changes have been done

**Comment by hpr-odoo** [comment]
Hi @aj-fuentes , let me know, if anything is remain here.
This is related to priority ticket so, I'm here again.

**Comment by aj-fuentes** [comment]
@hpr-odoo https://github.com/odoo/upgrade-util/pull/56

**Comment by hpr-odoo** [comment]
As #56 is merged

**Pull request closed without merging**